mypy: Final push to py3 annotations in core & enforce via linter

neiljp opened the pull request:
This concludes (almost) all the final changes from py2 to py3 annotation of functions, enforcing them through a linter addition.

One outstanding set of changes are to `cache.py` due to the comments in that file which suggest potential issues with such a transition; I thought it best to get what is there out so the linter can be enabled, if the changes as they are look good.

Otherwise, the linting is also currently disabled for scripts, tools, puppet and zerver/tests.

This passes `run-mypy` and `lint --backend` on my droplet.

timabbott commented:
@neiljp how did you do this migration?  Was it manual or did you use a tool?  Since this is a big refactor, I'm obviously worried about the possibility of bugs being introduced.

neiljp commented:
@timabbott Since most of the work was done during GCI and before, I made these changes manually.

Would you prefer an automated migration?

I have dropped (and added to the linter excludes) the changes to `api_test_helpers.py`, two changes involving `QuerySet[...]`, `api_test_helpers.py`, all but one change from `views/streams.py` (so fully excluded from linter) and the tornado descriptors changes. Also, I removed the zthumbor changes except for removing the ignore on the generic, adding minimal explicit typing instead, and switching `Any`->`Text` where seemed reasonable (though this commit could be dropped here, as well as `views/streams.py`, if it made it simpler to review.)

The remaining changes pass tests now; I hoped that lint and mypy would be sufficient, but this seems far from the case.

timabbott commented:
Manual is fine for this scale; I just wanted to know exactly what you did, since that impacts how I review :).  We probably will want automated if we're doing a bulk update to 100s of lines, which might be there for the tests.

The changes in `zerver/lib/request.py` look wrong to me; I think it might be important that we keep the `HttpRequest` first argument type passed through.  So I dropped that hunk and merged this.

Thanks @neiljp!

neiljp commented:
@timabbott My understanding is that the changes in request.py are actually largely cosmetic, since it's really `request.pyi` that is important here, which uses the format as suggested in my original commit.

If we do want to reintroduce this, we could consider the 'extended callable' notation which mypy offers?

timabbott commented:
Ahh, OK, that explains what I was missing :).  Still, I think it's probably important to have the `HttpRequest` type be declared explicitly wherever we annotate that function.
